# **Deep Learning for Safer Roads: Exploring CNN-Based and YOLOv8 Driver Drowsiness Detection**

This notebook delves into the application of advanced deep learning techniques to tackle a critical challenge in road safety: detecting driver drowsiness in real-time. By leveraging the **Driver Drowsiness Dataset (DDD)**, this project explores two promising approaches: **CNN-based models with Transfer Learning** and **YOLOv8 integrated with Roboflow**. Both techniques offer unique advantages, and this analysis aims to compare their effectiveness in building an accurate and efficient system for preventing fatigue-related accidents.

## **Introduction**
Driver fatigue is a significant contributor to road accidents worldwide, with drowsiness impairing reaction times and decision-making capabilities. Detecting such states in real-time provides a proactive approach to mitigating these risks. The **Driver Drowsiness Dataset (DDD)** offers a rich collection of RGB images of drivers, categorized into **Drowsy** and **Non-Drowsy** states. This dataset was created by extracting facial regions from real-life driving videos using the Viola-Jones algorithm, followed by preprocessing into images suitable for deep learning models.

In this project, we will examine two different methodologies for driver drowsiness detection: **CNN-based models with Transfer Learning** (such as MobileNetV2) and **YOLOv8 with Roboflow**. Each approach will be evaluated for its accuracy, efficiency, and real-time deployment potential.

## **Convolutional Neural Networks (CNNs) for Image Analysis**

**Convolutional Neural Networks (CNNs)** are a cornerstone in modern image recognition due to their ability to automatically learn spatial hierarchies of features from raw image data. CNNs are particularly effective for tasks such as facial state detection, which makes them ideal for identifying drowsy versus alert states in drivers. Key components of CNNs include:
1. **Convolution Layers**: These layers automatically extract essential features from the input images, such as edges and textures.
2. **Pooling Layers**: By reducing the spatial dimensions of the data, pooling helps improve computational efficiency and prevent overfitting.
3. **Fully Connected Layers**: These layers map the learned features to output labels, enabling the classification task.

CNNs have proven to be highly effective in many image-based tasks, including object detection and facial recognition, which makes them a great fit for analyzing driver facial states in this project.

## **MobileNetV2: An Optimized Solution for Real-Time Applications**

**MobileNetV2** is a modern convolutional architecture optimized for use in resource-constrained environments, such as mobile and embedded devices. This architecture provides a significant advantage in real-time applications, especially for systems deployed in vehicles. Key features that make MobileNetV2 suitable for this project include:
1. **Depthwise Separable Convolutions**: These reduce the computational cost by breaking down standard convolutions into simpler operations, improving efficiency while maintaining performance.
2. **Inverted Residuals with Linear Bottlenecks**: This design enhances information flow and reduces the model’s parameter count, making it suitable for real-time processing.
3. **Compact and Efficient Design**: MobileNetV2 delivers high accuracy while maintaining a small model size, making it ideal for applications where speed and low computational overhead are critical.

The lightweight design of MobileNetV2 makes it a strong candidate for deployment in real-time driver monitoring systems, where computational efficiency is key.

## **YOLOv8 with Roboflow for Real-Time Object Detection**

**YOLOv8 (You Only Look Once)** is a state-of-the-art object detection model that offers unparalleled speed and accuracy, making it perfect for real-time applications. When integrated with **Roboflow**, a platform that simplifies dataset preparation and model deployment, YOLOv8 becomes even more powerful for the task of detecting drowsiness in drivers. Benefits of using YOLOv8 with Roboflow include:
1. **Real-Time Detection and Localization**: YOLOv8 excels in both detecting and localizing objects, such as facial features, in real-time, making it ideal for driver monitoring.
2. **Comprehensive Object Detection**: YOLOv8 is capable of detecting multiple objects within a single frame, providing detailed insights into the driver's state.
3. **Streamlined Workflow**: Roboflow simplifies data annotation, augmentation, and deployment, offering an efficient pipeline for real-world applications.

This approach is particularly useful for applications that require fast and accurate detection in dynamic, real-time environments like vehicles.

## **Dataset Overview**
The **Driver Drowsiness Dataset (DDD)** serves as the foundation for training and testing both models:
- **Image Format**: RGB images with detailed facial features.
- **Classes**: **Drowsy** and **Non-Drowsy**.
- **Resolution**: 227 x 227 pixels, optimized for deep learning tasks.
- **Dataset Size**: Over 41,790 labeled images.
- **File Size**: Approximately 2.32 GB.

This dataset has been widely used in research, such as in the paper *“Detection and Prediction of Driver Drowsiness for the Prevention of Road Accidents Using Deep Neural Networks Techniques”* (DOI: [10.1007/978-981-33-6893-4_6](https://doi.org/10.1007/978-981-33-6893-4_6)).

## **Project Objectives**
The primary objectives of this project are:
1. **Evaluate both approaches**—CNN-based transfer learning models and YOLOv8 with Roboflow—for their ability to detect drowsy drivers.
2. **Compare performance** using metrics like accuracy, precision, recall, and F1-score.
3. **Optimize the models for real-time deployment**, ensuring they can be effectively integrated into vehicle monitoring systems.

## **Methodology**
This project will be carried out in several phases:
1. **Data Preprocessing**:
   - Normalize pixel values and augment data to enhance model generalization.
   - Split the dataset into training, validation, and testing sets.
2. **Model Development**:
   - Train a baseline CNN and implement **MobileNetV2** for transfer learning.
   - Train **YOLOv8** using Roboflow for real-time object detection and localization.
3. **Evaluation and Comparison**:
   - Assess the performance of each model using key metrics.
   - Compare the models' strengths in terms of speed, accuracy, and real-time feasibility.
4. **Real-Time Testing**:
   - Simulate real-world conditions to test the models’ responsiveness and deployment potential.

## **Significance**
This project represents a significant step toward improving road safety by leveraging AI to detect driver fatigue. By evaluating and comparing different deep learning approaches, this research aims to determine the most effective model for preventing fatigue-related accidents, ultimately contributing to safer roads for all.

Let’s dive into these innovative approaches and explore their potential for enhancing road safety!



## **Comparing Approaches: CNN + Transfer Learning vs YOLOv8 with Roboflow**

When tackling driver drowsiness detection, two advanced approaches emerge as key contenders: **CNN with Transfer Learning** and **YOLOv8 combined with Roboflow**. Below is a comprehensive comparison of their strengths and weaknesses for this critical task.

### **1. CNN + Transfer Learning: Simplicity Meets Power**

This method leverages pre-trained convolutional neural networks (e.g., MobileNetV2) for classifying driver states as drowsy or non-drowsy based on facial images.

#### **Why Choose CNN + Transfer Learning?**
- **Straightforward Setup**: Quick to implement and fine-tune using tools like TensorFlow or PyTorch.
- **Pre-Trained Knowledge**: Reduces training time by building on existing models.
- **Effective for Small Data**: Delivers solid performance even with limited datasets.
- **Customizable**: Tailor the final layers to your specific classification needs.

#### **The Drawbacks**
- **No Object Localization**: Focuses only on classification, not detecting multiple features in an image.
- **Real-Time Limitations**: Requires preprocessing (e.g., cropping faces), which adds latency.
- **Scalability Issues**: Performance may drop when handling high-resolution data or large datasets.

### **2. YOLOv8 + Roboflow: Real-Time Detection on Steroids**

YOLOv8 (You Only Look Once) is a state-of-the-art object detection model designed for speed, and Roboflow enhances its capabilities with seamless data preparation and deployment.

#### **Why YOLOv8 + Roboflow Stands Out**
- **Lightning-Fast Performance**: Perfect for real-time applications with minimal delay.
- **All-in-One Solution**: Detects and classifies in a single pipeline, skipping the need for separate preprocessing.
- **Robust Localization**: Pinpoints multiple objects (e.g., faces, eyes, or contextual features) simultaneously.
- **Streamlined Workflow**: Roboflow simplifies dataset annotation, augmentation, and deployment.

#### **Potential Pitfalls**
- **Resource Heavy**: Demands significant computational power for training and real-time use.
- **Higher Complexity**: Requires more expertise to fine-tune, especially for custom datasets.
- **Time-Intensive Annotations**: Needs a well-labeled dataset for accurate detection.

### **Key Differences: A Side-by-Side Comparison**

| **Aspect**                | **CNN + Transfer Learning**                     | **YOLOv8 + Roboflow**                         |
|---------------------------|------------------------------------------------|----------------------------------------------|
| **Primary Goal**          | Classification of drowsy vs. non-drowsy states | Real-time detection and localization         |
| **Real-Time Suitability** | Limited due to preprocessing latency           | Highly optimized for real-time applications  |
| **Ease of Use**           | Beginner-friendly and quick to set up          | More complex but streamlined with Roboflow   |
| **Dataset Needs**         | Performs well with smaller datasets            | Requires detailed annotations and larger datasets |
| **Performance Focus**     | High classification accuracy                   | Balances speed and detection accuracy        |
| **Hardware Requirements** | Moderate (suitable for CPUs or mid-tier GPUs)  | High (best for powerful GPUs)                |

---

### **Which Approach Should You Choose?**

- Opt for **CNN + Transfer Learning** if:
  - You’re focused on image classification without requiring object localization.
  - Your dataset is small and lacks detailed annotations.
  - You’re working with limited hardware or don’t need real-time detection.

- Go with **YOLOv8 + Roboflow** if:
  - Real-time detection and localization are essential for your application.
  - You have access to robust hardware and a well-annotated dataset.
  - You need an end-to-end system that combines detection and classification seamlessly.

---

Both methods are powerful tools for driver drowsiness detection, but the best choice depends on your project’s goals, constraints, and available resources.


## **Leveraging Transfer Learning with MobileNetV2 for CNN-Based Driver Drowsiness Detection**


## **Tools and Libraries Overview**

This project integrates various tools and libraries to streamline data processing, model training, and evaluation:

- **os**: Handles file system operations, such as managing dataset directories and paths.
- **pandas** and **numpy**: Facilitate data manipulation, exploration, and numerical computations essential for preprocessing and analysis.
- **matplotlib.pyplot** and **seaborn**: Enable clear and insightful data visualization, from basic plots to advanced statistical graphics like heatmaps.

For the machine learning pipeline, **TensorFlow** and **Keras** are used to design and train deep learning models. These frameworks provide powerful tools to build and customize architectures, such as **MobileNetV2**, a lightweight CNN optimized for efficiency and suitable for real-time applications. Additional layers, like **Dense** and **Flatten**, are included for feature extraction and classification.

To enhance training, optimizers such as **Adam** and **SGD** are employed for efficient weight updates. The **ImageDataGenerator** tool helps preprocess images and apply augmentations, improving the model's generalization ability.

Finally, performance evaluation relies on metrics like **confusion_matrix** and **classification_report** to assess accuracy, precision, recall, and F1-score. Utilities like **Counter** analyze class distributions, while **tf.random.set_seed** ensures reproducibility of results across experiments.

These components work together to build a reliable driver drowsiness detection system.



In [1]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as  plt
import tensorflow as tf
from tensorflow import keras
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
import seaborn as sns
tf.random.set_seed(3)
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.losses import SparseCategoricalCrossentropy
from tensorflow.keras.applications import MobileNetV2
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam, SGD
from collections import Counter

## **Dataset Download and File Count Verification**

This section automates the process of downloading a dataset from **Kaggle** and verifies its contents by counting the total number of files.

1. **Dataset Download**:
   - The `kagglehub` library is used to download the dataset with a single command.
   - The function `kagglehub.dataset_download("ismailnasri20/driver-drowsiness-dataset-ddd")` retrieves the dataset specified by its unique identifier on Kaggle.
   - The path to the downloaded dataset is returned and printed for reference.

2. **File Count Verification**:
   - The `os.walk` function iterates through all directories and subdirectories of the dataset folder.
   - Using `sum(len(files) for _, _, files in os.walk(dataset_path))`, the total number of files in the dataset is calculated and displayed.

- **Path to Dataset Files**: The directory where the dataset is stored after download.
- **Number of Files**: A quick verification step to ensure the dataset was successfully downloaded and contains the expected number of files.

This streamlined process saves time and ensures the dataset is readily accessible for the next steps in the project pipeline.

In [5]:
!pip install kaggle

In [15]:
import kagglehub
import os

# Download the dataset
dataset_path = kagglehub.dataset_download("ismailnasri20/driver-drowsiness-dataset-ddd")
print("Path to dataset files:", dataset_path)

# Count the number of files in the directory
file_count = sum(len(files) for _, _, files in os.walk(dataset_path))
print("Number of files in the directory:", file_count)

Path to dataset files: /root/.cache/kagglehub/datasets/ismailnasri20/driver-drowsiness-dataset-ddd/versions/1
Number of files in the directory: 41793


## **Dataset Splitting: Preparing Training, Validation, and Test Sets**

To ensure an effective training process, the dataset is divided into three subsets: **training**, **validation**, and **test**. This step is crucial for training the model, tuning hyperparameters, and evaluating its performance on unseen data.

Using the `splitfolders` library, the dataset is automatically split into the desired proportions:
- **Training Set (80%)**: Used to train the model and optimize its parameters.
- **Validation Set (15%)**: Helps fine-tune the model and prevent overfitting by monitoring performance on unseen data during training.
- **Test Set (5%)**: Used for the final evaluation of the model’s performance.

### Code Breakdown:
- `data_dir`: Specifies the path to the original dataset folder.
- `output_dir`: Defines the directory where the split datasets will be saved.
- `seed=1337`: Ensures reproducibility by setting a fixed random seed.
- `ratio=(0.8, 0.15, 0.05)`: Specifies the proportions for the splits (80% training, 15% validation, 5% test).


In [10]:
!pip install split-folders

In [22]:
for root, dirs, files in os.walk(dataset_path):
    print("Directory:", root)
    print("Contains directories:", dirs)
    print("Contains files:", files)

Directory: /root/.cache/kagglehub/datasets/ismailnasri20/driver-drowsiness-dataset-ddd/versions/1
Contains directories: ['Driver Drowsiness Dataset (DDD)']
Contains files: []
Directory: /root/.cache/kagglehub/datasets/ismailnasri20/driver-drowsiness-dataset-ddd/versions/1/Driver Drowsiness Dataset (DDD)
Contains directories: ['Drowsy', 'Non Drowsy']
Contains files: []
Directory: /root/.cache/kagglehub/datasets/ismailnasri20/driver-drowsiness-dataset-ddd/versions/1/Driver Drowsiness Dataset (DDD)/Drowsy
Contains directories: []
Contains files: ['B0182.png', 'ZB0312.png', 'O0317.png', 'O0214.png', 'T0907.png', 'C0104.png', 'P0769.png', 'B0299.png', 'I0642.png', 'J0285.png', 'ZA0940.png', 'ZA0996.png', 'ZB1245.png', 'L0855.png', 'W0663.png', 'A0584.png', 'T0655.png', 'Y0559.png', 'O0287.png', 'G0471.png', 'P0871.png', 'ZC1150.png', 'C0188.png', 'J0321.png', 'L0322.png', 'P0229.png', 'S0026.png', 'ZB0768.png', 'X0211.png', 'A1184.png', 'P0750.png', 'ZC0646.png', 'R0091.png', 'V0063.png', '

In [24]:
import splitfolders
import os

# Path to the subdirectory containing 'Drowsy' and 'Non Drowsy'
actual_dataset_path = os.path.join(
    dataset_path,
    "Driver Drowsiness Dataset (DDD)"
)

# Verify that the path contains subdirectories
print("Contents of the dataset directory:")
print(os.listdir(actual_dataset_path))

# Split the dataset
output_dir = 'splitted_dataset'
splitfolders.ratio(actual_dataset_path, output=output_dir, seed=1337, ratio=(.8, 0.15, 0.05))

Contents of the dataset directory:
['Drowsy', 'Non Drowsy']


Copying files: 41793 files [00:53, 777.60 files/s] 


By automating this process with `splitfolders.ratio`, the dataset is efficiently organized into separate folders, ready for use in the subsequent training pipeline.

## **Defining Dataset Directories**

This step specifies the paths to the dataset subsets, ensuring a clear and organized structure for the training pipeline.

1. **Directories**:
   - `train_dir`: Path to the **training dataset**. This subset contains the majority of the data (typically ~80%) and is used to train the model.
   - `test_dir`: Path to the **test dataset**. This subset (usually ~5%) is reserved for evaluating the model’s performance after training.
   - `val_dir`: Path to the **validation dataset**. This subset (~15%) is used during training to fine-tune the model and monitor its performance on unseen data.

2. **Purpose**:
   - These paths correspond to the output folders generated during the dataset splitting process (e.g., using `splitfolders`).
   - By explicitly defining these directories, the training pipeline can easily access the data needed for each stage of model development.


In [25]:
train_dir = "splitted_dataset/train"
test_dir = "splitted_dataset/test"
val_dir = "splitted_dataset/val"

This organization ensures a structured and reproducible workflow for preparing and training the deep learning model.

## **Data Normalization with ImageDataGenerator**

This step ensures that the image data is properly scaled, a crucial preprocessing task for deep learning models to improve convergence and performance.

### Code Breakdown:
1. **ImageDataGenerator**:
   - `ImageDataGenerator` is used to apply real-time data augmentation and preprocessing. In this case, it is primarily used for **rescaling** pixel values.
   
2. **Rescaling**:
   - `rescale=1./255`: This normalizes the pixel values of the images by scaling them from the range [0, 255] (typical for images) to [0, 1]. This ensures that the neural network trains efficiently, as neural networks generally perform better when input data is normalized.

3. **Different Sets**:
   - `train_datagen`, `test_datagen`, and `val_datagen` are created for the **training**, **testing**, and **validation** datasets, respectively. Each set undergoes the same preprocessing step (rescaling), ensuring consistency across the entire pipeline.


In [26]:
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)
val_datagen = ImageDataGenerator(rescale=1./255)

Normalization helps speed up the training process and leads to better performance by ensuring that the network receives input data in a consistent range.

## **Loading and Preprocessing the Training Data**

This step loads the training images from the specified directory and applies the necessary preprocessing for the model.

1. **flow_from_directory**:
   - `train_datagen.flow_from_directory` is used to load images directly from the directory (`train_dir`) and apply the preprocessing steps defined earlier (such as rescaling).

2. **Parameters**:
   - `target_size=(224, 224)`: This resizes all images to a consistent size of **224x224 pixels**, which is the input size expected by the model.
   - `batch_size=16`: The number of images processed at a time. A batch size of 16 allows the model to process 16 images simultaneously before updating its weights.
   - `class_mode='binary'`: Since the task is binary classification (drowsy vs non-drowsy), the labels are binary.
   - `shuffle=True`: Randomizes the order of images, which helps **prevent the model from learning unintended patterns** based on the order of data.

- This function prepares and loads the training images in batches, applying the necessary preprocessing, and provides them to the model for training. By using `flow_from_directory`, it efficiently handles large datasets without needing to manually load images into memory.


In [27]:
train_batches = train_datagen.flow_from_directory(
    train_dir,
    target_size=(224, 224),
    batch_size=16,
    class_mode='binary',
    shuffle=True
)

Found 33434 images belonging to 2 classes.


## **Loading and Preprocessing the Test Data**

This step is similar to the training data loading, but it is applied to the test dataset to evaluate the model's performance.

### Code Breakdown:
1. **flow_from_directory**:
   - `test_datagen.flow_from_directory` is used to load the images from the test directory (`test_dir`) and apply the preprocessing steps defined earlier (such as rescaling).

2. **Parameters**:
   - `target_size=(224, 224)`: All test images are resized to **224x224 pixels**, ensuring uniform input size for the model.
   - `batch_size=16`: The batch size of 16 means that 16 images will be processed in parallel during evaluation.
   - `class_mode='binary'`: Since the classification is binary (drowsy vs non-drowsy), labels are assigned as 0 or 1.
   - `shuffle=True`: The test data is shuffled, ensuring that the order does not bias the evaluation results.

- Similar to the training dataset, this function prepares and loads the test data in batches. It applies the necessary preprocessing to ensure the model receives consistent data for evaluation. By using `flow_from_directory`, it enables efficient handling of large test datasets without manually loading all images into memory.


In [28]:
test_batches = test_datagen.flow_from_directory(
    test_dir,
    target_size=(224, 224),
    batch_size=16,
    class_mode='binary' ,
    shuffle=True
)

Found 2091 images belonging to 2 classes.


## **Loading and Preprocessing the Validation Data**

This step prepares the validation dataset, ensuring that the model is evaluated on unseen data during the training process to prevent overfitting.

1. **flow_from_directory**:
   - `val_datagen.flow_from_directory` is used to load the images from the validation directory (`val_dir`) and apply the preprocessing steps defined earlier (such as rescaling).

2. **Parameters**:
   - `target_size=(224, 224)`: Resizes all validation images to **224x224 pixels**, matching the input size expected by the model.
   - `batch_size=16`: Loads and processes the images in batches of 16.
   - `class_mode='binary'`: Since the task is binary classification (drowsy vs non-drowsy), the labels are binary (0 or 1).
   - `shuffle=True`: Shuffles the validation dataset to ensure that the evaluation is not biased by the order of the images.

- This function loads and prepares the validation data in batches, applying the necessary preprocessing steps. The validation dataset is used during the training process to monitor the model's performance on unseen data, helping to fine-tune hyperparameters and detect overfitting.


In [29]:
val_batches = val_datagen.flow_from_directory(
    val_dir,
    target_size=(224, 224),
    batch_size=16,
    class_mode='binary',
    shuffle=True
)

Found 6268 images belonging to 2 classes.


## **Class Indices and Distribution Analysis**

This section extracts and analyzes the distribution of class labels (Drowsy vs. Non Drowsy) across the training, testing, and validation datasets.

1. **Class Indices**:
   - `train_class_indices`, `test_class_indices`, and `val_class_indices` store the indices associated with each class (e.g., 'Drowsy' and 'Non Drowsy') for the respective datasets. These indices help map the class names to numerical labels (0 and 1).

2. **Class Labels**:
   - `train_class_labels`, `test_class_labels`, and `val_class_labels` contain the actual class labels for each image in the datasets. These labels are used to compute the frequency of each class in the datasets.

3. **Class Counts**:
   - `train_class_counts`, `test_class_counts`, and `val_class_counts` are computed using the `Counter` class from the `collections` module. This provides a summary of the class distribution in each dataset.


In [30]:
train_class_indices = train_batches.class_indices
test_class_indices = test_batches.class_indices
val_class_indices = val_batches.class_indices

train_class_labels = train_batches.classes
test_class_labels = test_batches.classes
val_class_labels = val_batches.classes

train_class_counts = Counter(train_class_labels)
test_class_counts = Counter(test_class_labels)
val_class_counts = Counter(val_class_labels)

print("Class Names for train:\n", train_class_indices)
print("Class Counts for train:\n", train_class_counts)
print(end='\n')

print("Class Names for test:\n", test_class_indices)
print("Class Counts for test:\n", test_class_counts)
print(end='\n')

print("Class Names for validation :\n", val_class_indices)
print("Class Counts for validation:\n", val_class_counts)

Class Names for train:
 {'Drowsy': 0, 'Non Drowsy': 1}
Class Counts for train:
 Counter({0: 17878, 1: 15556})

Class Names for test:
 {'Drowsy': 0, 'Non Drowsy': 1}
Class Counts for test:
 Counter({0: 1118, 1: 973})

Class Names for validation :
 {'Drowsy': 0, 'Non Drowsy': 1}
Class Counts for validation:
 Counter({0: 3352, 1: 2916})


### Output Analysis:
- **Training Data**:
  - Class Names: `{'Drowsy': 0, 'Non Drowsy': 1}`
  - Class Counts: `{0: 17878, 1: 15556}`
  - Interpretation: The training dataset contains 17,878 images labeled as 'Drowsy' and 15,556 images labeled as 'Non Drowsy'. **The dataset is fairly balanced**, with a slightly higher number of drowsy images.

- **Testing Data**:
  - Class Names: `{'Drowsy': 0, 'Non Drowsy': 1}`
  - Class Counts: `{0: 1118, 1: 973}`
  - Interpretation: The testing dataset has 1,118 'Drowsy' images and 973 'Non Drowsy' images. While the numbers are smaller compared to the training set, the class distribution remains fairly consistent.

- **Validation Data**:
  - Class Names: `{'Drowsy': 0, 'Non Drowsy': 1}`
  - Class Counts: `{0: 3352, 1: 2916}`
  - Interpretation: The validation dataset has 3,352 'Drowsy' images and 2,916 'Non Drowsy' images. The class distribution is again balanced, although the drowsy class slightly outweighs the non-drowsy class.

These results provide insight into the class distribution across the datasets, which is important for model evaluation. A balanced distribution ensures that the model **does not learn biased patterns** and can generalize well to both classes. The class counts help identify potential class imbalances or overrepresented categories that may require attention during model training.